roachtest: asyncpg failed

roachtest.asyncpg [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3993762&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3993762&tab=artifacts#/asyncpg) on master @ [8c1528fff5c132b971428bc7f18bf0ea5eb16799](https://github.com/cockroachdb/cockroach/commits/8c1528fff5c132b971428bc7f18bf0ea5eb16799):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/asyncpg/run_1
	orm_helpers.go:245,orm_helpers.go:171,asyncpg.go:130,asyncpg.go:141,test_runner.go:780: 
		Tests run on Cockroach v22.1.0-alpha.00000000-2441-g8c1528fff5
		Tests run against asyncpg v0.24.0
		131 Total Tests Run
		129 tests passed
		2 tests failed
		34 tests skipped
		96 tests ignored
		0 tests passed unexpectedly
		2 tests failed unexpectedly
		0 tests expected failed but skipped
		0 tests expected failed but not run
		---
		--- FAIL: test_pool.TestPool.test_pool_connection_execute_many (unexpected)
		--- FAIL: test_execute.TestExecuteMany.test_executemany_basic (unexpected)
		For a full summary look at the asyncpg artifacts 
		
		An updated blocklist (asyncpgBlocklist22_1) is available in the artifacts' asyncpg log
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-experience
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*asyncpg.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.asyncpg [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3999330&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3999330&tab=artifacts#/asyncpg) on master @ [f3710ae65e327d0795123029db679f8c9f3e610e](https://github.com/cockroachdb/cockroach/commits/f3710ae65e327d0795123029db679f8c9f3e610e):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/asyncpg/run_1
	orm_helpers.go:245,orm_helpers.go:171,asyncpg.go:130,asyncpg.go:141,test_runner.go:780: 
		Tests run on Cockroach v22.1.0-alpha.00000000-2458-gf3710ae65e
		Tests run against asyncpg v0.24.0
		131 Total Tests Run
		129 tests passed
		2 tests failed
		34 tests skipped
		96 tests ignored
		0 tests passed unexpectedly
		2 tests failed unexpectedly
		0 tests expected failed but skipped
		0 tests expected failed but not run
		---
		--- FAIL: test_execute.TestExecuteMany.test_executemany_basic (unexpected)
		--- FAIL: test_pool.TestPool.test_pool_connection_execute_many (unexpected)
		For a full summary look at the asyncpg artifacts 
		
		An updated blocklist (asyncpgBlocklist22_1) is available in the artifacts' asyncpg log
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*asyncpg.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.asyncpg [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4004523&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4004523&tab=artifacts#/asyncpg) on master @ [01c0c739dd83b3e662d715404f2193d747f4fb9e](https://github.com/cockroachdb/cockroach/commits/01c0c739dd83b3e662d715404f2193d747f4fb9e):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/asyncpg/run_1
	orm_helpers.go:245,orm_helpers.go:171,asyncpg.go:130,asyncpg.go:141,test_runner.go:780: 
		Tests run on Cockroach v22.1.0-alpha.00000000-2464-g01c0c739dd
		Tests run against asyncpg v0.24.0
		131 Total Tests Run
		129 tests passed
		2 tests failed
		34 tests skipped
		96 tests ignored
		0 tests passed unexpectedly
		2 tests failed unexpectedly
		0 tests expected failed but skipped
		0 tests expected failed but not run
		---
		--- FAIL: test_execute.TestExecuteMany.test_executemany_basic (unexpected)
		--- FAIL: test_pool.TestPool.test_pool_connection_execute_many (unexpected)
		For a full summary look at the asyncpg artifacts 
		
		An updated blocklist (asyncpgBlocklist22_1) is available in the artifacts' asyncpg log
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*asyncpg.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.asyncpg [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4010617&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4010617&tab=artifacts#/asyncpg) on master @ [5ac733bb4927020bc1c52da24b2591742fde8e1f](https://github.com/cockroachdb/cockroach/commits/5ac733bb4927020bc1c52da24b2591742fde8e1f):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/asyncpg/run_1
	orm_helpers.go:245,orm_helpers.go:171,asyncpg.go:130,asyncpg.go:141,test_runner.go:780: 
		Tests run on Cockroach v22.1.0-alpha.00000000-2469-g5ac733bb49
		Tests run against asyncpg v0.24.0
		131 Total Tests Run
		129 tests passed
		2 tests failed
		34 tests skipped
		97 tests ignored
		0 tests passed unexpectedly
		2 tests failed unexpectedly
		0 tests expected failed but skipped
		0 tests expected failed but not run
		---
		--- FAIL: test_pool.TestPool.test_pool_connection_execute_many (unexpected)
		--- FAIL: test_execute.TestExecuteMany.test_executemany_basic (unexpected)
		For a full summary look at the asyncpg artifacts 
		
		An updated blocklist (asyncpgBlocklist22_1) is available in the artifacts' asyncpg log
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*asyncpg.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.asyncpg [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4017235&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4017235&tab=artifacts#/asyncpg) on master @ [3471dc50db8406f3faa638940492dbe0cab53435](https://github.com/cockroachdb/cockroach/commits/3471dc50db8406f3faa638940492dbe0cab53435):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/asyncpg/run_1
	orm_helpers.go:245,orm_helpers.go:171,asyncpg.go:130,asyncpg.go:141,test_runner.go:780: 
		Tests run on Cockroach v22.1.0-alpha.00000000-2505-g3471dc50db
		Tests run against asyncpg v0.24.0
		131 Total Tests Run
		129 tests passed
		2 tests failed
		34 tests skipped
		96 tests ignored
		0 tests passed unexpectedly
		2 tests failed unexpectedly
		0 tests expected failed but skipped
		0 tests expected failed but not run
		---
		--- FAIL: test_execute.TestExecuteMany.test_executemany_basic (unexpected)
		--- FAIL: test_pool.TestPool.test_pool_connection_execute_many (unexpected)
		For a full summary look at the asyncpg artifacts 
		
		An updated blocklist (asyncpgBlocklist22_1) is available in the artifacts' asyncpg log
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*asyncpg.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.asyncpg [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4022204&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4022204&tab=artifacts#/asyncpg) on master @ [fa9480a8b661b31a3c5e4c3428595b3e7c320700](https://github.com/cockroachdb/cockroach/commits/fa9480a8b661b31a3c5e4c3428595b3e7c320700):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/asyncpg/run_1
	orm_helpers.go:245,orm_helpers.go:171,asyncpg.go:130,asyncpg.go:141,test_runner.go:780: 
		Tests run on Cockroach v22.1.0-alpha.00000000-2509-gfa9480a8b6
		Tests run against asyncpg v0.24.0
		131 Total Tests Run
		129 tests passed
		2 tests failed
		34 tests skipped
		96 tests ignored
		0 tests passed unexpectedly
		2 tests failed unexpectedly
		0 tests expected failed but skipped
		0 tests expected failed but not run
		---
		--- FAIL: test_pool.TestPool.test_pool_connection_execute_many (unexpected)
		--- FAIL: test_execute.TestExecuteMany.test_executemany_basic (unexpected)
		For a full summary look at the asyncpg artifacts 
		
		An updated blocklist (asyncpgBlocklist22_1) is available in the artifacts' asyncpg log
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*asyncpg.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.asyncpg [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4026851&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4026851&tab=artifacts#/asyncpg) on master @ [65c6377f4f486d537ef7dfc59f8f9c78f1a6018a](https://github.com/cockroachdb/cockroach/commits/65c6377f4f486d537ef7dfc59f8f9c78f1a6018a):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/asyncpg/run_1
	orm_helpers.go:245,orm_helpers.go:171,asyncpg.go:130,asyncpg.go:141,test_runner.go:780: 
		Tests run on Cockroach v22.1.0-alpha.00000000-2511-g65c6377f4f
		Tests run against asyncpg v0.24.0
		131 Total Tests Run
		129 tests passed
		2 tests failed
		34 tests skipped
		96 tests ignored
		0 tests passed unexpectedly
		2 tests failed unexpectedly
		0 tests expected failed but skipped
		0 tests expected failed but not run
		---
		--- FAIL: test_pool.TestPool.test_pool_connection_execute_many (unexpected)
		--- FAIL: test_execute.TestExecuteMany.test_executemany_basic (unexpected)
		For a full summary look at the asyncpg artifacts 
		
		An updated blocklist (asyncpgBlocklist22_1) is available in the artifacts' asyncpg log
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*asyncpg.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.asyncpg [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4033018&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4033018&tab=artifacts#/asyncpg) on master @ [e3be114fc6fe8eda07aafbc7ee43cc452456f270](https://github.com/cockroachdb/cockroach/commits/e3be114fc6fe8eda07aafbc7ee43cc452456f270):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/asyncpg/run_1
	orm_helpers.go:245,orm_helpers.go:171,asyncpg.go:130,asyncpg.go:141,test_runner.go:780: 
		Tests run on Cockroach v22.1.0-alpha.00000000-2548-ge3be114fc6
		Tests run against asyncpg v0.24.0
		131 Total Tests Run
		129 tests passed
		2 tests failed
		34 tests skipped
		96 tests ignored
		0 tests passed unexpectedly
		2 tests failed unexpectedly
		0 tests expected failed but skipped
		0 tests expected failed but not run
		---
		--- FAIL: test_pool.TestPool.test_pool_connection_execute_many (unexpected)
		--- FAIL: test_execute.TestExecuteMany.test_executemany_basic (unexpected)
		For a full summary look at the asyncpg artifacts 
		
		An updated blocklist (asyncpgBlocklist22_1) is available in the artifacts' asyncpg log
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*asyncpg.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/server/purge_auth_session.go
+++ b/pkg/server/purge_auth_session.go
@@ -124,7 +124,7 @@ RETURNING 1
 		autoLogoutTime = currTime.Add(autoLogoutTimeout * time.Duration(-1))
 	)
 
-	if _, err := internalExecutor.QueryRowEx(
+	if _, err := internalExecutor.ExecEx(
 		ctx,
 		"delete-old-expired-sessions",
 		nil, /* txn */
@@ -136,7 +136,7 @@ RETURNING 1
 		log.Errorf(ctx, "error while deleting old expired web sessions: %+v", err)
 	}
 
-	if _, err := internalExecutor.QueryRowEx(
+	if _, err := internalExecutor.ExecEx(
 		ctx,
 		"delete-old-revoked-sessions",
 		nil, /* txn */
@@ -148,7 +148,7 @@ RETURNING 1
 		log.Errorf(ctx, "error while deleting old revoked web sessions: %+v", err)
 	}
 
-	if _, err := internalExecutor.QueryRowEx(
+	if _, err := internalExecutor.ExecEx(
 		ctx,
 		"delete-sessions-timeout",
 		nil, /* txn */

--- a/pkg/sql/conn_executor_exec.go
+++ b/pkg/sql/conn_executor_exec.go
@@ -694,7 +694,12 @@ func (ex *connExecutor) execStmtInOpenState(
 	p.extendedEvalCtx.Annotations = &p.semaCtx.Annotations
 	p.stmt = stmt
 	p.cancelChecker.Reset(ctx)
-	p.autoCommit = os.ImplicitTxn.Get() && !ex.server.cfg.TestingKnobs.DisableAutoCommit
+
+	// We need to turn off autocommit behavior here so that the "insert fast path"
+	// does not get triggered. The postgres docs say that commands in the extended
+	// protocol are all treated as an implicit transaction that does not get
+	// committed until a Sync message is received.
+	p.autoCommit = os.ImplicitTxn.Get() && !isExtendedProtocol && !ex.server.cfg.TestingKnobs.DisableAutoCommit
 
 	var stmtThresholdSpan *tracing.Span
 	alreadyRecording := ex.transitionCtx.sessionTracing.Enabled()

--- a/pkg/sql/pgwire/testdata/pgtest/portals
+++ b/pkg/sql/pgwire/testdata/pgtest/portals
@@ -1048,3 +1048,98 @@ ReadyForQuery
 {"Type":"DataRow","Values":[{"text":"1"}]}
 {"Type":"CommandComplete","CommandTag":"SELECT 1"}
 {"Type":"ReadyForQuery","TxStatus":"I"}
+
+send
+Query {"String": "DROP TABLE IF EXISTS portal_sync_test"}
+----
+
+until ignore=NoticeResponse
+ReadyForQuery
+----
+{"Type":"CommandComplete","CommandTag":"DROP TABLE"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+send
+Query {"String": "CREATE TABLE portal_sync_test (a FLOAT CHECK(a > 1))"}
+----
+
+until
+ReadyForQuery
+----
+{"Type":"CommandComplete","CommandTag":"CREATE TABLE"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+send
+Parse {"Name": "s15", "Query": "INSERT INTO portal_sync_test VALUES($1);"}
+Describe {"ObjectType": "S", "Name": "s15"}
+Sync
+----
+
+until
+ReadyForQuery
+----
+{"Type":"ParseComplete"}
+{"Type":"ParameterDescription","ParameterOIDs":[701]}
+{"Type":"NoData"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+# Verify that the "insert fast path" does not cause an auto-commit before
+# the Sync message is handled. This also tests behavior with the unnamed
+# portal.
+send
+Bind {"PreparedStatement": "s15", "Parameters": [{"text":"2"}]}
+Describe {"ObjectType": "P"}
+Execute
+Bind {"PreparedStatement": "s15", "Parameters": [{"text":"3"}]}
+Describe {"ObjectType": "P"}
+Execute
+Sync
+----
+
+until
+ReadyForQuery
+----
+{"Type":"BindComplete"}
+{"Type":"NoData"}
+{"Type":"CommandComplete","CommandTag":"INSERT 0 1"}
+{"Type":"BindComplete"}
+{"Type":"NoData"}
+{"Type":"CommandComplete","CommandTag":"INSERT 0 1"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+# Verify that an error during the implicit transaction causes both
+# INSERTs to be rolled back.
+send
+Bind {"PreparedStatement": "s15", "Parameters": [{"text":"4"}]}
+Describe {"ObjectType": "P"}
+Execute
+Bind {"PreparedStatement": "s15", "Parameters": [{"text":"0"}]}
+Describe {"ObjectType": "P"}
+Execute
+Sync
+----
+
+until ignore_constraint_name
+ErrorResponse
+ReadyForQuery
+----
+{"Type":"BindComplete"}
+{"Type":"NoData"}
+{"Type":"CommandComplete","CommandTag":"INSERT 0 1"}
+{"Type":"BindComplete"}
+{"Type":"NoData"}
+{"Type":"ErrorResponse","Code":"23514"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+send
+Query {"String": "SELECT * FROM portal_sync_test ORDER BY a;"}
+----
+
+until ignore_table_oids
+ReadyForQuery
+----
+{"Type":"RowDescription","Fields":[{"Name":"a","TableOID":0,"TableAttributeNumber":1,"DataTypeOID":701,"DataTypeSize":8,"TypeModifier":-1,"Format":0}]}
+{"Type":"DataRow","Values":[{"text":"2"}]}
+{"Type":"DataRow","Values":[{"text":"3"}]}
+{"Type":"CommandComplete","CommandTag":"SELECT 2"}
+{"Type":"ReadyForQuery","TxStatus":"I"}

--- a/pkg/testutils/pgtest/datadriven.go
+++ b/pkg/testutils/pgtest/datadriven.go
@@ -181,6 +181,12 @@ func MsgsToJSONWithIgnore(msgs []pgproto3.BackendMessage, args *datadriven.TestD
 					}
 				}
 			}
+		case "ignore_constraint_name":
+			for _, msg := range msgs {
+				if m, ok := msg.(*pgproto3.ErrorResponse); ok {
+					m.ConstraintName = ""
+				}
+			}
 		case "ignore":
 			for _, typ := range arg.Vals {
 				ignore[fmt.Sprintf("*pgproto3.%s", typ)] = true

